docs/dev.rst

```diff
@@ -0,0 +1,53 @@
+:orphan:
+
+.. _dev:
+
+=======================
+Developer Documentation
+=======================
+
+This documentation summarizes how the AutoPyTorch code works, and it is meant to guide developers
+on how to best contribute to it.
+
+AutoPyTorch relies on the `SMAC <https://automl.github.io/SMAC3/master/>`_ library to build individual models,
+which are later ensembled together using ensemble selection by `Caruana et al. (2004) <https://dl.acm.org/doi/pdf/10.1145/1015330.1015432>`_.
+Therefore, there are two main parts of the code: `AutoMLSMBO`, which is our interface to the SMAC package, and
+`EnsembleBuilder` which opportunistically builds an ensemble of the individual algorithms found by SMAC, at fixed intervals.
+The following sections provides details regarding this two main blocks of code.
+
+Building Individual Models
+==========================
+
+AutoPyTorch relies on Scikit-Learn `Pipeline <https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html>`_ to build an individual algorithm.
+In other words, each of the individual models fitted by SMAC are (and comply) with Scikit-Learn pipeline and framework. For example, when a pipeline is fitted,
```

Comment:
A pipeline can consist of various preprocessing steps including- imputation, encoding, scaling, and feature preprocessing for tabular data followed by the setup of the training algorithm which includes building a network embedding for categorical features, a network backbone that is responsible for feature extraction, a network head that is responsible for decision making and choosing an appropriate Learning Rate scheduler, optimisation algorithm, and initialisation of the network parameters. 

The preprocessing steps are aggregated into a Sklearn ColumnTransformer to maintain the order of the columns and the column transformer is wrapped into a TabularColumnTransformer class which complies with a torchvision transform. Based on the size of the dataset, we apply this transformer to the whole of the dataset in the EarlyPreprocessing node if the dataset is small enough or it is added as a transform to the data loader object that allows transforming the dataset batchwise.

Reply:
This can also be moved to the end (AutoML part) where you talk about the pipeline